Implement a keep-alive feature for directly controlled heater switches

pdcastro opened the pull request:
The heater switch keep-alive feature consists of regularly refreshing the state of directly controlled switches at a configurable interval (regularly turning the switch ‘on’ or ‘off’ _again,_ even if it is already turned ‘on’ or ‘off’), just like the `keep_alive` setting of Home Assistant’s Generic Thermostat integration:

> HASS Generic Thermostat https://www.home-assistant.io/integrations/generic_thermostat/
> `keep_alive` _integer (optional)_
> _Set a keep-alive interval. If set, the switch specified in the heater option will be triggered every time the interval elapses. Use with heaters and A/C units that shut off if they don’t receive a signal from their remote for a while. Use also with switches that might lose state._

pdcastro commented:
@jmcollin78, this PR proposes the implementation of a keep-alive feature as described. If you agree to having such a feature added as proposed in this PR generally, I can then work further on it, especially adding documentation and translations.

When rebased on tag `5.3.3.beta3`, all tests pass in the VSCode dev container. I have also tested “the real thing” with an appliance of mine and Home Assistant version 2024.1.3. The code was linted with black.

pdcastro commented on custom_components/versatile_thermostat/underlyings.py:
```diff
@@ -202,6 +204,7 @@ def __init__(
         self._on_time_sec = 0
         self._off_time_sec = 0
         self._hvac_mode = None
+        self._keep_alive = IntervalCaller(hass, keep_alive_sec)
```
Note: To simplify the code, I chose to always instantiate this object even if `keep_alive_sec` is unset or zero, but in this case it will do nothing and the feature will be effectively disabled. I tested it when `keep_alive_sec` is unset, zero, and non-zero.

pdcastro commented on custom_components/versatile_thermostat/underlyings.py:
```diff
@@ -214,6 +217,13 @@ def is_inversed(self):
+        super().startup()
+        self._keep_alive.set_async_action(
+            self.turn_on if self.is_device_active else self.turn_off
+        )
```
As per other comment, even though this piece of code always runs, it will do nothing if the `keep_alive` setting is unset or zero.

jmcollin78 replied:
I  have seen after that it is done below. You can forget this comment.

jmcollin78 commented:
Oooh I will see that. Thank you for this PR !

For over_switch, because of the cycle which constantly turns on and off, I guess there is no need of keep_alive (that is why I don't implement it initialy). And don't know for over_climate if this make sens, because the underlying climate may do it alone and may be for `over_valve` ?

I'm not sure this is useful finaly. In what case do you need this feature ?

pdcastro commented:
> Oooh I will see that. Thank you for this PR !
> 
> For over_switch, because of the cycle which constantly turns on and off, I guess there is no need of keep_alive (that is why I don't implement it initialy). And don't know for over_climate if this make sens, because the underlying climate may do it alone and may be for `over_valve` ?
> 
> I'm not sure this is useful finaly. In what case do you need this feature ?

Thanks for the feedback. The keep-alive feature is intended for “over_switch” operation — directly controlled switch entities. When a TPI cycle is delivering high power, say 100% (when the target temperature is well above the current temperature and the “on_percent” variable is near 1 = 100%), I understand that the switch is turned ‘on’ at the beginning of the cycle, and no further action (switch on/off) is taken on the switch until the end of the cycle. Does this match your understanding as well? In this case, if the appliance requires a keep-alive interval that is shorter than the TPI cycle, then the appliance will switch off (failsafe mode) before the end of the TPI cycle, delivering less power than intended.

I have definitely observed this behaviour with a small electric heater that requires a short keep-alive interval of 20 seconds, with a TPI cycle of 1 minute and intended 100% power. Without the keep-alive feature implemented in this PR, the appliance would switch off about 20 seconds into the TPI cycle, thus delivering 33% power (20 seconds ‘on’, 40 seconds ‘off’) even when the intended TPI power was 100% (60 seconds 'on', 0 seconds 'off').

Reducing the TPI cycle duration to match the required keep-alive interval may work in some cases, but not all cases. A larger heating system based on oil or natural gas (currently over three quarters of homes here in the UK, for example) may require longer TPI cycles such as 10 or 20 minutes,* while switch keep-alive intervals might be 1 or 5 minutes (or any other value, depending on the system design).

\* Example: My central heating system takes up to a minute between heat demand being presented and the boiler igniting the flame, so a TPI cycle of 1 minute would just not work. When heat demand is presented to the system, a mechanical water valve slowly opens and some 20 seconds later, a water pump turns on heat demand is presented to a natural gas boiler. The boiler then turns on a fan and takes another 20 to 40 seconds to ignite the main flame. I find that 20 minutes is a reasonable TPI cycle for this system.

jmcollin78 commented:
> When a TPI cycle is delivering high power, say 100% (when the target temperature is well above the current temperature and the “on_percent” variable is near 1 = 100%), I understand that the switch is turned ‘on’ at the beginning of the cycle, and no further action (switch on/off) is taken on the switch until the end of the cycle

Yes, you are totally right. When 100% or 0% the switch is not switched but only at start of the first cycle. THe use case is now clear for me. I will have a look before going further to the development of the release (based on beta already pubished).

jmcollin78 commented on tests/test_start.py:
```diff
@@ -264,6 +264,7 @@ async def test_over_switch_deactivate_preset(
             CONF_HEATER_2: None,
             CONF_HEATER_3: None,
             CONF_HEATER_4: None,
+            CONF_HEATER_KEEP_ALIVE: 0,
```
Please add a unit test with KEEP_ALIVE not 0 to see if method are called.

You can inspire from the existing tests `test_bug_66` which mocks method, check they are called and waits for the window to be open/close.

Let me know if it is too complex I'm pretty sure I can help.

jmcollin78 reviewed (commented):
Great improvement @pdcastro !

I will take this PR with pleasure. The comments are easy to fix, I guess.

pdcastro reviewed (commented):
Thanks for you review. 👍  I am adding new commits that address the points that were raised, to make it easier for you to see what has changed. Before merging the PR though, I could squash the commits to make the history tidier.
I also still need to add documentation and translations (Google Translate...). I can add those to this same PR, or I can create a separate PR. Let me know if you have a preference.

pdcastro commented on custom_components/versatile_thermostat/underlyings.py:
```diff
@@ -245,12 +255,13 @@ async def turn_off(self):
-                command,
-                data,
-            )
+            try:
+                data = {ATTR_ENTITY_ID: self._entity_id}
+                await self._hass.services.async_call(domain, command, data)
+                self._keep_alive.set_async_action(self.turn_off)
+            except Exception:
+                self._keep_alive.cancel()
+                raise
```
Building on previous comments, I’ve added an inner `try` block that catches and re-raises any exception for the purpose of cancelling keep alive if the service call fails for any reason. I had the following cases in mind:
- Keep alive initiated through the `startup()` method, before it is known whether the service calls will succeed.
- Keep alive initiated here, such that at least one service call succeeded, but where subsequent service calls might fail (for any unforeseen reason, perhaps in case the user renames or removes the service while Versatile Thermostat is running).

Adding the inner try/except block allows a single `cancel()` call to be made. An alternative might have been to make two `cancel()` calls:

```python
        try:
            data = {ATTR_ENTITY_ID: self._entity_id}
            await self._hass.services.async_call(domain, command, data)
            self._keep_alive.set_async_action(self.turn_on)
        except ServiceNotFound as err:
            _LOGGER.error(err)
            self._keep_alive.cancel()
        except Exception:
            self._keep_alive.cancel()
            raise
```

By the way:
- `pylint` doesn’t seem to complain about “broad exception caught” when the exception is re-raised, which is why I did not need to add `# pylint: disable=broad-exception-caught`.
- `black` was happy with the re-formatting.

pdcastro commented on custom_components/versatile_thermostat/keep_alive.py:
```diff
@@ -0,0 +1,53 @@
+        self.cancel()
+
+        async def callback(_time: datetime):
+            try:
+                await action()
+            except Exception as e:  # pylint: disable=broad-exception-caught
+                _LOGGER.error(e)
+                self.cancel()
```
The added try/except block avoids ugly job/task stack traces printed in the Home Assistant logs, which would look like a bug in the Versatile Thermostat integration. For example, a raising an unexpected “foo” exception _without_ the try/except block:
```
2024-01-23 23:50:21.545 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved:   File "/home/vscode/.local/bin/hass", line 8, in <module>
    sys.exit(main())
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/runner.py", line 188, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 640, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 607, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 1914, in _run_once
    handle._run()
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/helpers/event.py", line 1454, in _run_async_call_action
    hass.async_run_hass_job(job, time_tracker_utcnow())
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/core.py", line 671, in async_run_hass_job
    hassjob.target(*args)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/helpers/event.py", line 1532, in interval_listener
    hass.async_run_hass_job(job, now)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/core.py", line 674, in async_run_hass_job
    return self.async_add_hass_job(hassjob, *args)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/core.py", line 568, in async_add_hass_job
    task = self.loop.create_task(hassjob.target(*args), name=hassjob.name)
Traceback (most recent call last):
  File "/workspaces/versatile_thermostat/custom_components/versatile_thermostat/keep_alive.py", line 47, in callback
    raise Exception("foo")
Exception: foo
```
And _with_ the try/except block:
```
2024-01-23 23:49:32.513 ERROR (MainThread) [custom_components.versatile_thermostat.keep_alive] foo
```

pdcastro reviewed (commented):
New commits added to address the review comments. To confirm:
- All tests pass (`pytest` in the VS Code dev container).
- Code formatted with `black`.
- Code checked with `pylint` and `pylance`.

pdcastro commented on custom_components/versatile_thermostat/thermostat_switch.py:
```diff
@@ -101,6 +102,7 @@ def post_init(self, config_entry):
                     thermostat=self,
                     switch_entity_id=switch,
                     initial_delay_sec=idx * delta_cycle,
+                    keep_alive_sec=config_entry.get(CONF_HEATER_KEEP_ALIVE),
```
Actually, if it is just `None` that we are guarding against, the following should be enough, no?
```python
UnderlyingSwitch(
    ... ,
    keep_alive_sec=config_entry.get(CONF_HEATER_KEEP_ALIVE, 0),
)
```
I did some testing and I can confirm that:
* `config_entry.get(SOME_UNSET_ENTRY, 0)` returns `0` rather than `None`.
* When using the Home Assistant integration configuration user interface (config flow), the UI prevents me from entering negative numbers and decimal points, and if I delete the contents of the form field (unsetting it), `config_entry.get(SOME_UNSET_ENTRY)` actually returns `0` rather than `None`, thanks to the schema constraint in `config_schema.py`:

```python
vol.Optional(CONF_HEATER_KEEP_ALIVE, default=0): cv.positive_int,
```

jmcollin78 replied:
I would do that:
```
keep_alive_sec=config_entry.get(CONF_HEATER_KEEP_ALIVE) or 0 
```
but 
```
config_entry.get(CONF_HEATER_KEEP_ALIVE, 0)
```
(which I don't was aware until now), seems good also.

pdcastro replied:
`config_entry` has the type of [ConfigEntry.data](https://github.com/home-assistant/core/blob/2024.1.5/homeassistant/config_entries.py#L242), which is [collections.abc.Mapping](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping), which is specified at:
https://github.com/python/cpython/blob/v3.11.7/Lib/_collections_abc.py#L771

While investigating this, I got a bit carried away and created a separate PR that adds more type hints in a few files: #364

jmcollin78 commented on custom_components/versatile_thermostat/thermostat_switch.py:
```diff
@@ -121,6 +123,7 @@ async def async_added_to_hass(self):
                     self.hass, [switch.entity_id], self._async_switch_changed
                 )
             )
+            switch.startup()
```
ok. Thank for the explanation.

> “Please note that keep-alive messages will only start after the thermostat is first switched on,”

So, if I understand well, in summer, when Vtherm is off, the keep-alive will continue to be send to the device ? Am I wrong ? Why not but I wonder how to avoid that. I consider permanently running features even if turned off is a bad pattern. But maybe I understand a wrong way.

jmcollin78 replied:
> “VTherm” is short for “Versatile Thermostat”, right?

Yes

jmcollin78 commented on custom_components/versatile_thermostat/keep_alive.py:
```diff
@@ -0,0 +1,43 @@
+            self._remove_handle = None
+
+    def set_async_action(self, action: Callable[[], Awaitable[None]]):
+        if not self._interval_sec:
```
Yes good point. Testing once for all before arriving here is better. I confirm.

---
In a general way, an int can always be None. Python is not a strongly typed language (like Javascript if you want). Typing is more for an indication for helping the user.

jmcollin78 commented on custom_components/versatile_thermostat/keep_alive.py:
```diff
@@ -0,0 +1,43 @@
+    def set_async_action(self, action: Callable[[], Awaitable[None]]):
+        if not self._interval_sec:
+            return
+        self.cancel()
```
Yes ... but ... for the future robustness of the code (many people as me will change this in the future) is it, in my opinion better to cancel() even if you think there is nothing to cancel.

This kind of programing "save my life" many time as I handles also non nominal cases. There is many piece of code like this (stop heater even if "normally" it is already turned off -> this is safety). Testing if xx <= 0 instead of if xxx == 0, ... all this patterns are for safety and robustness.

In that case, in the cancel method, if really there is nothing to cancel, then you do nothing. So, the cancel() method have to be think as "cancel_if_any", if you prefer.

pdcastro replied:
> I will do it if you insist further!

You haven’t yet insisted further, 🫣 so I am leaving it as it is. I understand that you are not quite happy with it, but I think I can live with that! 😄😇 The code is good.

jmcollin78 commented:
> Thanks for you review. 👍 I am adding new commits that address the points that were raised, to make it easier for you to see what has changed. Before merging the PR though, I could squash the commits to make the history tidier. I also still need to add documentation and translations (Google Translate...). I can add those to this same PR, or I can create a separate PR. Let me know if you have a preference.

I prefer to have all in the same PR.

pdcastro commented on README-fr.md:
```diff
@@ -244,7 +244,10 @@ En fonction de votre choix sur le type de thermostat, vous devrez choisir une ou
-![image](https://github.com/jmcollin78/versatile_thermostat/blob/main/images/config-linked-entity.png?raw=true)
+![image](images/config-linked-entity.png)
+
+Certains switches de radiateur nécessitent des « messages keep-alive » pour les empêcher de déclencher un arrêt de sécurité. Ceci peut être activé en entrant un nombre de secondes non nul dans le champ d'intervalle keep-alive du switch. En cas de doute, laissez-le vide ou zéro (valeur par défaut).
```
Pardon my French! And blame Google Translate for it. 🙂

pdcastro commented on custom_components/versatile_thermostat/translations/fr.json:
```diff
@@ -46,6 +46,7 @@
           "heater_entity2_id": "2ème radiateur",
           "heater_entity3_id": "3ème radiateur",
           "heater_entity4_id": "4ème radiateur",
+          "heater_keep_alive": "Intervalle keep-alive du switch en secondes",
```
I saw that the word “switch” was already being used elsewhere in the French translation, so I used it here too instead of “interrupteur”, but you tell me! As for “keep-alive”, I wonder if it is a technical term that is more readily understood in English than when translated. Again, you tell me. Google Translate actually suggested some strange things equivalent to _“Change the interval of life maintenance”._ 😖

pdcastro commented on custom_components/versatile_thermostat/translations/el.json:
```diff
@@ -50,6 +51,7 @@
           "heater_entity2_id": "Προαιρετική 2η ταυτότητα οντότητας θερμαντήρα. Αφήστε κενό αν δεν χρησιμοποιείται",
           "heater_entity3_id": "Προαιρετική 3η ταυτότητα οντότητας θερμαντήρα. Αφήστε κενό αν δεν χρησιμοποιείται",
           "heater_entity4_id": "Προαιρετική 4η ταυτότητα οντότητας θερμαντήρα. Αφήστε κενό αν δεν χρησιμοποιείται",
+          "heater_keep_alive": "Optional heater switch state refresh interval. Disabled if empty or zero",
```
I had a go at French and Italian because I had some hope of being able to make absurd Google translations slightly less absurd, but Greek and Slovak I will not even try. A copy-and-paste of Google Translate would end up as some nonsense like _“Option to replace the country’s freshen up intermission”_ (or worse).

So I entered English as placeholder for the Greek and Slovak translations. I suggest that we merge this PR (when you feel like it is tolerable), and straight after that we invite **`misa1515`, `vassilis-papanikolaou`** and **`Bergoglio`** to provide translations (either they create a new PR, or we do). We could also put this PR on hold until they provide translations, but this PR already has 40+ comments, touches 21 files, and if they took a while to reply, the PR could end up with merge conflicts.

jmcollin78 replied:
No. HA will do itself the fallback to the En version. Just don't add the values for language you don't know.

Some contributors make a PR regularly to adapt the text. I will ping them after the release so that they will be able to add their texte.

pdcastro commented:
PR status: I have added the documentation and translations for review, with some comments. Other than that, I think the PR is ready for merging. I have squashed two of the commits too (tidier history). I have not marked any conversations as “resolved” — I leave that to you as the reviewer. If you think I have missed any conversation, please point it out to me. Thanks!

jmcollin78 commented on custom_components/versatile_thermostat/translations/fr.json:
```diff
@@ -67,6 +68,7 @@
           "heater_entity2_id": "Optionnel entity id du 2ème radiateur",
           "heater_entity3_id": "Optionnel entity id du 3ème radiateur",
           "heater_entity4_id": "Optionnel entity id du 4ème radiateur",
+          "heater_keep_alive": "Optionnel intervalle de rafraîchissement du switch. Désactivé si vide ou zéro",
```
"Intervalle de rafraichissement du switch en secondes. Mettre 0 ou vide pour désactiver. À n'utiliser que pour les switchs qui le nécessite."

jmcollin78 commented on custom_components/versatile_thermostat/translations/en.json:
```diff
@@ -302,6 +305,7 @@
           "heater_entity2_id": "Optional 2nd Heater entity id. Leave empty if not used",
           "heater_entity3_id": "Optional 3rd Heater entity id. Leave empty if not used",
           "heater_entity4_id": "Optional 4th Heater entity id. Leave empty if not used",
+          "heater_keep_alive": "Optional heater switch state refresh interval. Disabled if empty or zero",
```
add "Use it only if your switch need it." at the end of the description. Idem for step/option description.

pdcastro replied:
I agree that it makes sense to advise the user that this feature should only be used when required, 👍 as it has “performance implications” of sorts. But I also noticed that the whole description was getting long:

```python
"heater_keep_alive": "Optional heater switch state refresh interval. Disabled if empty or zero. Use it only if your switch needs it."
```

 The preceding descriptions, for example, have a more succinct _“Leave empty if not used”_ instruction:
```python
"heater_entity2_id": "Optional 2nd Heater entity id. Leave empty if not used",
"heater_entity3_id": "Optional 3rd Heater entity id. Leave empty if not used",
"heater_entity4_id": "Optional 4th Heater entity id. Leave empty if not used",
```

I had added _"Disabled if empty or zero”_ because my schema definition meant that the form was pre-filled with `0`, so the user needed to be advised of the meaning of `0`:

```python
vol.Optional(CONF_HEATER_KEEP_ALIVE, default=0): cv.positive_int,
```

But from the end user’s point of view, leaving the form empty is more straightforward than entering zero. So I've changed the schema for the default to be an empty field:

```python
vol.Optional(CONF_HEATER_KEEP_ALIVE): cv.positive_int,
```

I have tested to confirm than an empty field doe not cause any validation issues with the `positive_int` constraint, and the Python code replaces `None` with `0` to satisfy the `int` type hint:

```python
 keep_alive_sec=config_entry.get(CONF_HEATER_KEEP_ALIVE, 0)
```

So, there’s no longer a need to mention “zero” to the user. My new proposal is to say _“Leave empty if not required”:_

```python
"heater_keep_alive": "Optional heater switch state refresh interval. Leave empty if not required.",
```

I think this also removes the need to say _”Use it only if your switch needs it.”_ The preceding `heater_entity` descriptions read _”Leave empty **if not used**”_ rather than _”**if not required**”_ and I think that’s appropriate: If you are not _”using”_ an extra switch entity then you leave the field empty _”if not used”,_ whereas if your switch does not _”require”_ keep-alive messages, then you leave the field empty _”if not required”._ Furthermore, while _”Disabled if empty or zero”_ was just a piece of information, _“Leave empty if not required”_ is a more forceful instruction of what the user should do.

Would you agree? If you agree, please let me have amended French translations as well. I will wait for your feedback before updating the screenshots.

jmcollin78 replied:
Great.

jmcollin78 commented on custom_components/versatile_thermostat/translations/sk.json:
```diff
@@ -45,6 +45,7 @@
           "heater_entity2_id": "2. spínač ohrievača",
           "heater_entity3_id": "3. spínač ohrievača",
           "heater_entity4_id": "4. spínač ohrievača",
+          "heater_keep_alive": "Switch keep-alive interval in seconds",
```
remove. Fallback is done by HA itself. Idem below. If you are sure for Italian, you can keep it.

jmcollin78 commented on tests/test_switch_keep_alive.py:
```diff
@@ -0,0 +1,256 @@
+                call("switch", SERVICE_TURN_OFF, {"entity_id": "switch.mock_switch"}),
+                call("switch", SERVICE_TURN_OFF, {"entity_id": "switch.mock_switch"}),
+            ],
+        )
```
Great test class ! I have learn something. Really nice.

jmcollin78 commented:
We are almost done. And you have now a conflict in readme since your PR in images. Sorry for this but it should be easy to fix.

This is really a great job you have done 😍

pdcastro commented on custom_components/versatile_thermostat/config_schema.py:
```diff
@@ -98,6 +98,7 @@
         vol.Optional(CONF_HEATER_4): selector.EntitySelector(
             selector.EntitySelectorConfig(domain=[SWITCH_DOMAIN, INPUT_BOOLEAN_DOMAIN]),
         ),
+        vol.Optional(CONF_HEATER_KEEP_ALIVE): cv.positive_int,
```
Amended to remove `default=0` as explained in https://github.com/jmcollin78/versatile_thermostat/pull/345#discussion_r1468544063

pdcastro commented on custom_components/versatile_thermostat/strings.json:
```diff
@@ -67,6 +68,7 @@
           "heater_entity2_id": "Optional 2nd Heater entity id. Leave empty if not used",
           "heater_entity3_id": "Optional 3rd Heater entity id. Leave empty if not used",
           "heater_entity4_id": "Optional 4th Heater entity id. Leave empty if not used",
+          "heater_keep_alive": "Optional heater switch state refresh interval. Leave empty if not required.",
```
See comment https://github.com/jmcollin78/versatile_thermostat/pull/345#discussion_r1468544063.
Also, regarding the final period punctuation `’.’` at the end of _”Leave empty if not required.”_ , I have noticed that it is not consistent with the preceding descriptions that to not include the final period, but I included it because:
* You included the period in your suggested English and French revisions, and
* I think including the period is the right thing to do when the sentence is “full” rather than just a “label.” For example, _”Optional heater switch”_ is just a label, whereas _“This is an optional heater switch.”_ and _”Leave empty if not required.”_ are full sentences. (And if a sentence follows a label, a period is needed at the end of the label in order to separate it from the sentence.)

pdcastro commented on README.md:
```diff
@@ -654,9 +657,9 @@ context:
+## Parameter summary
 
-| Paramètre                                 | Libellé                                                                       | "over switch" | "over climate"      | "over valve" | "central configuration" |
+| Parameter                                 | Description                                                                   | "over switch" | "over climate"      | "over valve" | "central configuration" |
```
Oh, and I have also amended this table since your last review.

jmcollin78 commented on README-fr.md:
````diff
@@ -245,6 +245,9 @@ En fonction de votre choix sur le type de thermostat, vous devrez choisir une ou
 ### Pour un thermostat de type ```thermostat_over_switch```
 ![image](images/config-linked-entity.png)
+
+Certains équipements nécessitent d'être périodiquement sollicités pour les empêcher un arrêt de sécurité. Connu sous le nom de "keep-alive" cette fonction est activable en entrant un nombre de secondes non nul dans le champ d'intervalle keep-alive du thermostat. Pour désactiver la fonction ou en cas de doute, laissez-le vide ou entrez zéro (valeur par défaut).
````
"Certains équipements nécessitent d'être périodiquement sollicités pour  empêcher un arrêt de sécurité": remove 'les'

jmcollin78 commented on custom_components/versatile_thermostat/translations/fr.json:
```diff
@@ -67,6 +68,7 @@
           "heater_entity2_id": "Optionnel entity id du 2ème radiateur",
           "heater_entity3_id": "Optionnel entity id du 3ème radiateur",
           "heater_entity4_id": "Optionnel entity id du 4ème radiateur",
+          "heater_keep_alive": "Intervalle de rafraichissement du switch en secondes. Mettre 0 ou vide pour désactiver. À n'utiliser que pour les switchs qui le nécessite.",
```
"Intervalle de rafraichissement du switch en secondes. Laisser vide pour désactiver. À n'utiliser que pour les switchs qui le nécessite."

jmcollin78 commented on custom_components/versatile_thermostat/translations/fr.json:
```diff
@@ -314,6 +317,7 @@
           "heater_entity2_id": "Optionnel entity id du 2ème radiateur",
           "heater_entity3_id": "Optionnel entity id du 3ème radiateur",
           "heater_entity4_id": "Optionnel entity id du 4ème radiateur",
+          "heater_keep_alive": "Intervalle de rafraichissement du switch en secondes. Mettre 0 ou vide pour désactiver. À n'utiliser que pour les switchs qui le nécessite.",
```
idem

pdcastro commented:
I think it’s all done! 🎉

jmcollin78 reviewed (approved):
Nice work !

jmcollin78 commented:
Merged !
